File: queue.cpp
#include <stdio.h>
#include <iostream>

using namespace std;

struct Queue {
  private:
    unsigned capacity;
    int front;
    int* items;
    unsigned int size = 0;
    int tail;

  public:
    Queue(unsigned int c): capacity(c) {
      this->items = new int[c];
    }

    void clear() {
      for (int i = 0; i < this->size; i++) {
        this->items[i] = NULL;

        for (int j = i; j < this->size; j++) {
          this->items[j] = this->items[j + 1];
        }
      }
    }

    int dequeue() {
      if (this->empty()) {
        return NULL;
      }

      for (int i = 0; i < this->size; i++) {
        this->items[i] = this->items[i + 1];
      }

      this->front = this->items[0];

      return this->front;
    }

    bool empty() {
      return this->size == 0;
    }

    void enqueue(int item) {
      if (this->full()) {
        throw runtime_error("[Queue]: is full!");
      }

      this->tail = item;
      this->items[this->size] = item;
      this->size++;
    }

    int front() {
      return this->front;
    }

    bool full() {
      return this->size == this->capacity;
    }

    void show() {
      for (int i = 0; i < this->size; i++) {
        cout << "Item: " << this->items[i] << " | Position: " << i << endl;
      }
    }

    int size() {
      return this->size;
    }

    int tail() {
      return this->tail;
    }
};

int main() {
  return 0;
}

File: go/queue/queue.go
package queue

import (
	"errors"
	"fmt"
)

type Queue struct {
	capacity, size uint64
	items          []*int64
	front, tail    *int64
}

func NewQueue(capacity uint64) *Queue {
	return &Queue{
		capacity: capacity,
		size:     0,
		items:    []*int64{},
		front:    nil,
		tail:     nil,
	}
}

func (q *Queue) clear() {
	var i uint64 = 0

	for i < (q.size - 1) {
		q.items[i] = nil
		i++
	}

	q.front = nil
	q.tail = nil
	q.size = 0
}

func (q *Queue) dequeue() {
	if q.empty() {
		return
	}

	var i uint64 = 1

	for i <= q.size {
		q.items[i-1] = q.items[i]
		i++
	}

	q.front = q.items[0]
	q.size--
}

func (q *Queue) enqueue(item *int64) error {
	if q.full() {
		return errors.New("[queue] is full")
	}

	q.items[q.size] = item
	q.tail = item
	q.size++

	return nil
}

func (q *Queue) empty() bool {
	return q.size == 0
}

func (q *Queue) full() bool {
	return q.size == q.capacity
}

func (q *Queue) show() {
	var i uint64 = 0

	for i < q.size {
		fmt.Println(q.items[i])
		i++
	}
}
